.config/doom/config.org
#+TITLE: SlimTux's Emacs Config
#+AUTHOR: SlimTux
#+SETUPFILE: https://fniessen.github.io/org-html-themes/org/theme-readtheorg.setup
#+EXPORT_FILE_NAME: index.html
#+HTML_HEAD: <style>pre.src{background:#343131;color:white;} </style>
#+OPTIONS: toc: t
* Introduction
This is an Elisp program that runs every time I start my Emacs editor. Some people also call it an *emacs config*. Why am I writing it under Org mode? /Literate Programming, Bitch/!

- *Author*: SlimTux
- *Created*: 06-10-2023
- *License*: [[./LICENSE][GNU General Public License (GPL)]]

#+CAPTION: XKCD Emacs Comic
#+NAME: fig:Emacs meme
[[https://imgs.xkcd.com/comics/real_programmers.png]]
* Package Management
Emacs is like an operating system; it is highly extensible. You can install new functionality by just installing a package, similar to how you do it in an OS. You can also configure these programs by changing some Elisp variables.

#+BEGIN_VERSE
An average Emacs user can easily end up installing 100+ packages.
#+END_VERSE

** Melpa
- *Reference*: https://www.melpa.org/

By default, Emacs contains a limited number of package choices. This is why we need Melpa. It's a package repository for Emacs.

#+BEGIN_VERSE
MELPA is an ELPA-compatible package repository that contains an enormous number of useful Emacs packages.
#+END_VERSE

Think of it like the VSCode Plugin Marketplace (eww).

#+BEGIN_SRC elisp
(require 'package)
(setq package-enable-at-startup nil)

(add-to-list 'package-archives
             '("melpa" . "https://melpa.org/packages/"))
(package-initialize)
#+END_SRC

* UI
** Fonts(FIX-ME)
#+begin_src emacs-lisp
(setq doom-font  (font-spec :family "JetBrains Mono" :size 17))
#+end_src
** Theme
#+begin_src emacs-lisp
(setq doom-theme 'doom-dracula)
#+end_src
** Line Numbers
#+begin_src emacs-lisp
(setq display-line-numbers-type t)
#+end_src
#+begin_src emacs-lisp
(use-package rainbow-mode
  :ensure t
  :init (add-hook 'prog-mode-hook 'rainbow-mode))
#+end_src
#+begin_src emacs-lisp
(setq display-time-default-load-average nil)
#+end_src
** Dashboard
#+begin_src emacs-lisp
(use-package dashboard
  :ensure t
  :config
    (dashboard-setup-startup-hook)
    (setq dashboard-startup-banner "~/.emacs.d/img/avatar.png")
    (setq dashboard-items '((recents  . 5)
			      (projects . 5)))
    (setq dashboard-banner-logo-title "I am just a coder for fun"))
#+end_src
* Meta
** Org Directory
#+begin_src emacs-lisp
(setq org-directory "~/org/")
#+end_src
#+begin_src emacs-lisp
(setq inhibit-startup-message t)
#+end_src
** Key-bidings
*** Visit the configuration
#+begin_src emacs-lisp
(defun config-visit ()
  (interactive)
  (find-file "~/.config/doom/config.org"))
(global-set-key (kbd "C-c e") 'config-visit)
#+end_src
*** Reload the configuration
#+begin_src emacs-lisp
(defun config-reload ()
  (interactive)
  (org-babel-load-file (expand-file-name "~/.config/doom/config.org")))
(global-set-key (kbd "C-c r") 'config-reload)
#+end_src
* Programs
*** Editing Features
**** Hungry Delete
- /Reference: https://github.com/nflath/hungry-delete/

Hungry Delete is a minor-mode that causes deletion to delete all whitespace in the direction you are deleting. Works exactly like c-hungry-delete-mode, which is where the code was from. This just packages it up to be easier to use in other modes.

#+BEGIN_SRC elisp
  (use-package hungry-delete
    :ensure t
    :defer t
    :config (global-hungry-delete-mode))
#+END_SRC
*** Browser Text
#+begin_src elisp
(use-package atomic-chrome
  :ensure t
  :init
  (setq atomic-chrome-default-major-mode 'markdown-mode)
  (setq atomic-chrome-server-ghost-text-port 4002)
  (setq atomic-chrome-extension-type-list '(atomic-chrome))
  :config
  (atomic-chrome-start-server))
#+end_src

*** Vterm
#+begin_src emacs-lisp
(use-package vterm
  :ensure t
  :init
  (global-set-key (kbd "<s-return>") 'vterm))
#+end_src
*** Fancy battery
#+begin_src emacs-lisp
(use-package fancy-battery
  :ensure t
  :init
  (fancy-battery-mode 1)
  (setq fancy-battery-show-percentage t))
#+end_src
*** Latex
**** AucTeX
#+begin_src emacs-lisp
(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq TeX-save-query nil)
;(setq TeX-PDF-mode t)

(require 'flymake)

(defun flymake-get-tex-args (file-name)
(list "pdflatex"
(list "-file-line-error" "-draftmode" "-interaction=nonstopmode" file-name)))

(add-hook 'LaTeX-mode-hook 'flymake-mode)

(setq ispell-program-name "aspell") ; could be ispell as well, depending on your preferences
(setq ispell-dictionary "english") ; this can obviously be set to any language your spell-checking program supports

(add-hook 'LaTeX-mode-hook 'flyspell-mode)
(add-hook 'LaTeX-mode-hook 'flyspell-buffer)

(defun turn-on-outline-minor-mode ()
(outline-minor-mode 1))

(add-hook 'LaTeX-mode-hook 'turn-on-outline-minor-mode)
(add-hook 'latex-mode-hook 'turn-on-outline-minor-mode)
(setq outline-minor-mode-prefix "\C-c \C-o") ; Or something else

(require 'tex-site)
(autoload 'reftex-mode "reftex" "RefTeX Minor Mode" t)
(autoload 'turn-on-reftex "reftex" "RefTeX Minor Mode" nil)
(autoload 'reftex-citation "reftex-cite" "Make citation" nil)
(autoload 'reftex-index-phrase-mode "reftex-index" "Phrase Mode" t)
(add-hook 'latex-mode-hook 'turn-on-reftex) ; with Emacs latex mode
;; (add-hook 'reftex-load-hook 'imenu-add-menubar-index)
(add-hook 'LaTeX-mode-hook 'turn-on-reftex)

(setq LaTeX-eqnarray-label "eq"
LaTeX-equation-label "eq"
LaTeX-figure-label "fig"
LaTeX-table-label "tab"
LaTeX-myChapter-label "chap"
TeX-auto-save t
TeX-newline-function 'reindent-then-newline-and-indent
TeX-parse-self t
TeX-style-path
'("style/" "auto/"
"/usr/share/emacs/site-lisp/auctex/style/")
LaTeX-section-hook
'(LaTeX-section-heading
LaTeX-section-title
LaTeX-section-toc
LaTeX-section-section
LaTeX-section-label))
#+end_src
** preview
#+begin_src elisp
(use-package org-latex-impatient
  :defer t
  :hook (org-mode . org-latex-impatient-mode)
  :init
  (setq org-latex-impatient-tex2svg-bin
        ;; location of tex2svg executable
        "~/node_modules/mathjax-node-cli/bin/tex2svg"))
#+end_src
** Auto Start
#+begin_src emacs-lisp
(use-package centaur-tabs
  :ensure t
  :init
  (centaur-tabs-mode 1))
#+end_src



*** Highlight Indent Guides
The "highlight-indent-guides" package is an Emacs extension that enhances code readability by providing visual indent guides. As you work with code, it displays vertical lines at each level of indentation, making it easier to distinguish different blocks and understand the code's structure. This feature is particularly useful for languages with significant indentation, such as Python. "highlight-indent-guides" helps maintain consistent and well-organized code, ensuring a more pleasant coding experience in Emacs.

- /Reference - https://github.com/DarthFennec/highlight-indent-guides/

#+begin_src elisp
  (use-package highlight-indent-guides
    :ensure t
    :defer t
    :hook (prog-mode . highlight-indent-guides-mode)
    :config
    (setq highlight-indent-guides-method 'character)
    (setq highlight-indent-guides-character ?\|)
    (setq highlight-indent-guides-responsive 'top))
#+end_src
** Emacs LSP
The "emacs-lsp" project is a collection of Emacs packages and tools that implement the Language Server Protocol (LSP) in Emacs. LSP is a standardized communication protocol that enables integration with language servers, which are external programs providing advanced code analysis and language-specific features.

The goal of the "emacs-lsp" project is to enhance the Emacs text editor and turn it into a powerful Integrated Development Environment (IDE) by leveraging the capabilities of language servers. These packages provide language-specific features such as autocompletion, real-time error checking, code navigation, and more. By adhering to the LSP, developers can use a consistent approach across various programming languages, streamlining their workflow and improving productivity.

The project offers a diverse range of packages, each tailored to specific programming languages and their corresponding language servers. This initiative fosters an integrated and standardized environment for Emacs users, enabling them to efficiently code in different languages and benefit from advanced language-specific tooling within their favorite text editor. The "emacs-lsp" project is a valuable resource for developers seeking a robust and unified coding experience in Emacs.

- /Reference - https://emacs-lsp.github.io/

*** LSP Mode
LSP mode (Language Server Protocol mode) in Emacs is a powerful extension that brings IDE-like capabilities to various programming languages. It provides integration with language servers, which are separate programs that offer advanced code analysis, autocompletion, and other language-specific features. LSP mode allows developers to benefit from a consistent development experience across different programming languages, eliminating the need for language-specific configurations and setups. With LSP mode, Emacs users can enjoy enhanced code navigation, error checking, and automatic code formatting, significantly improving their productivity and coding efficiency.

#+CAPTION: LSP mode completion
#+NAME: fig:LSP completion
https://emacs-lsp.github.io/lsp-mode/examples/completion.gif

- /Reference - https://github.com/emacs-lsp/lsp-mode/

#+begin_src elisp
  (use-package lsp-mode
    :ensure t
    :defer t
    :init
    (setq lsp-keymap-prefix "C-c l")
    :config
    (setq lsp-headerline-breadcrumb-enable nil))
#+end_src
** Languages
Now, we will configure language server protocol and other settings for each,
languages I work on, one by one.
*** Web (html/css/js)
**** Web Mode
Web Mode in Emacs is a major mode that enhances web development by providing specialized editing features for working with HTML, CSS, JavaScript, and other web-related languages. It intelligently handles nested tags, auto-closes HTML tags, and offers indentation and syntax highlighting tailored for web development. Web Mode also supports embedded templates and server-side code, making it a versatile tool for web developers to efficiently create and edit web pages and applications within the Emacs text editor.

- /Reference: https://web-mode.org/

#+begin_src elisp
  (use-package web-mode
    :ensure t
    :defer t
    :config
    (setq
     web-mode-markup-indent-offset 2
     web-mode-css-indent-offset 2
     web-mode-code-indent-offset 2
     web-mode-style-padding 2
     web-mode-script-padding 2
     web-mode-enable-auto-closing t
     web-mode-enable-auto-opening t
     web-mode-enable-auto-pairing t
     web-mode-enable-auto-indentation t)
    :mode
    (".html$" "*.php$" "*.tsx"))
#+end_src

**** Emmet-mode
Emmet Mode in Emacs is an extension that significantly boosts web development productivity by enabling advanced HTML and CSS abbreviations. Originally inspired by the Emmet toolkit, this mode allows developers to write complex markup with ease using intuitive shortcuts and expand them into full HTML or CSS code. It supports dynamic placeholders, numeric repetition, and custom abbreviation expansion, making it a powerful tool for rapidly generating and editing HTML and CSS structures. With Emmet Mode, Emacs users can streamline their web development workflow, saving time and effort while maintaining clean and well-structured code.

- /Reference: https://https://github.com/smihica/emmet-mode/

#+CAPTION: Emmet Mode Demo
#+NAME:   fig:emmet mode
[[https://www.philnewton.net/assets/blog/2015/08/emmet.gif]]

#+begin_src elisp
  (use-package emmet-mode
    :ensure t
    :defer t)
#+end_src
* Org Mode
Org Mode in Emacs is a powerful and versatile outlining and organizing tool that extends the text editor's capabilities beyond simple plain text. It provides a structured and hierarchical format for creating notes, to-do lists, project plans, and more. Org Mode offers features such as headings, lists, tables, tags, and timestamps, enabling users to manage complex information with ease. It supports exporting to various formats like HTML, PDF, and LaTeX, making it suitable for both personal organization and professional document preparation. With its extensive functionality and seamless integration with Emacs, Org Mode empowers users to efficiently manage tasks, maintain documentation, and stay organized in a clutter-free and efficient environment.

** Org Bullet
Org Bullet is an Emacs package that enhances the visual appearance of Org Mode outlines by replacing plain text bullet points with custom symbols. It offers a variety of stylish bullets to represent different outline levels, making the organization of tasks and information more visually appealing and easier to comprehend. Org Bullet is highly configurable, allowing users to customize the bullet symbols to their preference and create a more visually pleasing and organized presentation of hierarchical data within Org Mode documents.

- /Reference - https://github.com/sabof/org-bullets/

#+begin_src elisp
  (use-package org-bullets
    :ensure t
    :defer t
    :config
    (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))
#+end_src

** Org Agenda
Org Agenda in Emacs is a powerful and flexible tool for managing tasks, appointments, and notes. It provides a dynamic view of scheduled events, deadlines, and TODO items from various Org Mode files, enabling users to organize and prioritize their work effectively. With its customizable views, filters, and sorting options, Org Agenda offers a comprehensive overview of upcoming events and pending tasks, making it an essential feature for staying organized and productive in Emacs.

#+begin_src elisp
  (setq org-agenda-files (append
                          (file-expand-wildcards "~/org/*.org")))
#+end_src

* Misc
Some extra setting, which doesn't fall in any category above.
** Locales
- /Reference: https://www.gnu.org/software/emacs/manual/html_node/elisp/Locales.html/
#+begin_src elisp
  (setq locale-coding-system 'utf-8)
  (set-terminal-coding-system 'utf-8)
  (set-keyboard-coding-system 'utf-8)
  (set-selection-coding-system 'utf-8)
  (prefer-coding-system 'utf-8)
#+end_src


.config/doom/config-backup.org
#+TITLE: SlimTux's Emacs Config
#+AUTHOR: SlimTux
#+SETUPFILE: https://fniessen.github.io/org-html-themes/org/theme-readtheorg.setup
#+EXPORT_FILE_NAME: index.html
* Introduction
This is an Elisp program that runs every time I start my Emacs editor. Some people also call it an *emacs config*. Why am I writing it under Org mode? /Literate Programming, Bitch/!

- *Author*: SlimTux
- *Created*: 06-10-2023
- *License*: [[./LICENSE][GNU General Public License (GPL)]]

#+CAPTION: XKCD Emacs Comic
#+NAME: fig:Emacs meme
[[https://imgs.xkcd.com/comics/real_programmers.png]]
* Package Management
Emacs is like an operating system; it is highly extensible. You can install new functionality by just installing a package, similar to how you do it in an OS. You can also configure these programs by changing some Elisp variables.

#+BEGIN_VERSE
An average Emacs user can easily end up installing 100+ packages.
#+END_VERSE

** Melpa
- *Reference*: https://www.melpa.org/

By default, Emacs contains a limited number of package choices. This is why we need Melpa. It's a package repository for Emacs.

#+BEGIN_VERSE
MELPA is an ELPA-compatible package repository that contains an enormous number of useful Emacs packages.
#+END_VERSE

Think of it like the VSCode Plugin Marketplace (eww).

#+BEGIN_SRC elisp
(require 'package)
(setq package-enable-at-startup nil)

(add-to-list 'package-archives
             '("melpa" . "https://melpa.org/packages/"))
(package-initialize)
#+END_SRC

* UI
** Fonts(FIX-ME)
#+begin_src emacs-lisp
(setq doom-font  (font-spec :family "JetBrains Mono" :size 17))
#+end_src
** Theme
#+begin_src emacs-lisp
(setq doom-theme 'doom-dracula)
#+end_src
** Line Numbers
#+begin_src emacs-lisp
(setq display-line-numbers-type t)
#+end_src
* Meta
** Org Directory
#+begin_src emacs-lisp
(setq org-directory "~/org/")
#+end_src
* Programs
*** Editing Features
**** Hungry Delete
- /Reference: https://github.com/nflath/hungry-delete/

Hungry Delete is a minor-mode that causes deletion to delete all whitespace in the direction you are deleting. Works exactly like c-hungry-delete-mode, which is where the code was from. This just packages it up to be easier to use in other modes.

#+BEGIN_SRC elisp
  (use-package hungry-delete
    :ensure t
    :defer t
    :config (global-hungry-delete-mode))
#+END_SRC
** Highlight Indent Guides
The "highlight-indent-guides" package is an Emacs extension that enhances code readability by providing visual indent guides. As you work with code, it displays vertical lines at each level of indentation, making it easier to distinguish different blocks and understand the code's structure. This feature is particularly useful for languages with significant indentation, such as Python. "highlight-indent-guides" helps maintain consistent and well-organized code, ensuring a more pleasant coding experience in Emacs.

- /Reference - https://github.com/DarthFennec/highlight-indent-guides/

#+begin_src elisp
  (use-package highlight-indent-guides
    :ensure t
    :defer t
    :hook (prog-mode . highlight-indent-guides-mode)
    :config
    (setq highlight-indent-guides-method 'character)
    (setq highlight-indent-guides-character ?\|)
    (setq highlight-indent-guides-responsive 'top))
#+end_src
** Emacs LSP
The "emacs-lsp" project is a collection of Emacs packages and tools that implement the Language Server Protocol (LSP) in Emacs. LSP is a standardized communication protocol that enables integration with language servers, which are external programs providing advanced code analysis and language-specific features.

The goal of the "emacs-lsp" project is to enhance the Emacs text editor and turn it into a powerful Integrated Development Environment (IDE) by leveraging the capabilities of language servers. These packages provide language-specific features such as autocompletion, real-time error checking, code navigation, and more. By adhering to the LSP, developers can use a consistent approach across various programming languages, streamlining their workflow and improving productivity.

The project offers a diverse range of packages, each tailored to specific programming languages and their corresponding language servers. This initiative fosters an integrated and standardized environment for Emacs users, enabling them to efficiently code in different languages and benefit from advanced language-specific tooling within their favorite text editor. The "emacs-lsp" project is a valuable resource for developers seeking a robust and unified coding experience in Emacs.

- /Reference - https://emacs-lsp.github.io/

*** LSP Mode
LSP mode (Language Server Protocol mode) in Emacs is a powerful extension that brings IDE-like capabilities to various programming languages. It provides integration with language servers, which are separate programs that offer advanced code analysis, autocompletion, and other language-specific features. LSP mode allows developers to benefit from a consistent development experience across different programming languages, eliminating the need for language-specific configurations and setups. With LSP mode, Emacs users can enjoy enhanced code navigation, error checking, and automatic code formatting, significantly improving their productivity and coding efficiency.

#+CAPTION: LSP mode completion
#+NAME: fig:LSP completion
https://emacs-lsp.github.io/lsp-mode/examples/completion.gif

- /Reference - https://github.com/emacs-lsp/lsp-mode/

#+begin_src elisp
  (use-package lsp-mode
    :ensure t
    :defer t
    :init
    (setq lsp-keymap-prefix "C-c l")
    :config
    (setq lsp-headerline-breadcrumb-enable nil))
#+end_src
** Languages
Now, we will configure language server protocol and other settings for each,
languages I work on, one by one.
*** Web (html/css/js)
**** Web Mode
Web Mode in Emacs is a major mode that enhances web development by providing specialized editing features for working with HTML, CSS, JavaScript, and other web-related languages. It intelligently handles nested tags, auto-closes HTML tags, and offers indentation and syntax highlighting tailored for web development. Web Mode also supports embedded templates and server-side code, making it a versatile tool for web developers to efficiently create and edit web pages and applications within the Emacs text editor.

- /Reference: https://web-mode.org/

#+begin_src elisp
  (use-package web-mode
    :ensure t
    :defer t
    :config
    (setq
     web-mode-markup-indent-offset 2
     web-mode-css-indent-offset 2
     web-mode-code-indent-offset 2
     web-mode-style-padding 2
     web-mode-script-padding 2
     web-mode-enable-auto-closing t
     web-mode-enable-auto-opening t
     web-mode-enable-auto-pairing t
     web-mode-enable-auto-indentation t)
    :mode
    (".html$" "*.php$" "*.tsx"))
#+end_src

**** Emmet-mode
Emmet Mode in Emacs is an extension that significantly boosts web development productivity by enabling advanced HTML and CSS abbreviations. Originally inspired by the Emmet toolkit, this mode allows developers to write complex markup with ease using intuitive shortcuts and expand them into full HTML or CSS code. It supports dynamic placeholders, numeric repetition, and custom abbreviation expansion, making it a powerful tool for rapidly generating and editing HTML and CSS structures. With Emmet Mode, Emacs users can streamline their web development workflow, saving time and effort while maintaining clean and well-structured code.

- /Reference: https://https://github.com/smihica/emmet-mode/

#+CAPTION: Emmet Mode Demo
#+NAME:   fig:emmet mode
[[https://www.philnewton.net/assets/blog/2015/08/emmet.gif]]

#+begin_src elisp
  (use-package emmet-mode
    :ensure t
    :defer t)
#+end_src
* Org Mode
Org Mode in Emacs is a powerful and versatile outlining and organizing tool that extends the text editor's capabilities beyond simple plain text. It provides a structured and hierarchical format for creating notes, to-do lists, project plans, and more. Org Mode offers features such as headings, lists, tables, tags, and timestamps, enabling users to manage complex information with ease. It supports exporting to various formats like HTML, PDF, and LaTeX, making it suitable for both personal organization and professional document preparation. With its extensive functionality and seamless integration with Emacs, Org Mode empowers users to efficiently manage tasks, maintain documentation, and stay organized in a clutter-free and efficient environment.

** Org Bullet
Org Bullet is an Emacs package that enhances the visual appearance of Org Mode outlines by replacing plain text bullet points with custom symbols. It offers a variety of stylish bullets to represent different outline levels, making the organization of tasks and information more visually appealing and easier to comprehend. Org Bullet is highly configurable, allowing users to customize the bullet symbols to their preference and create a more visually pleasing and organized presentation of hierarchical data within Org Mode documents.

- /Reference - https://github.com/sabof/org-bullets/

#+begin_src elisp
  (use-package org-bullets
    :ensure t
    :defer t
    :config
    (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))
#+end_src

** Org Agenda
Org Agenda in Emacs is a powerful and flexible tool for managing tasks, appointments, and notes. It provides a dynamic view of scheduled events, deadlines, and TODO items from various Org Mode files, enabling users to organize and prioritize their work effectively. With its customizable views, filters, and sorting options, Org Agenda offers a comprehensive overview of upcoming events and pending tasks, making it an essential feature for staying organized and productive in Emacs.

#+begin_src elisp
  (setq org-agenda-files (append
                          (file-expand-wildcards "~/dox/org/*.org")))
#+end_src

* Misc
Some extra setting, which doesn't fall in any category above.
** Locales
- /Reference: https://www.gnu.org/software/emacs/manual/html_node/elisp/Locales.html/
#+begin_src elisp
  (setq locale-coding-system 'utf-8)
  (set-terminal-coding-system 'utf-8)
  (set-keyboard-coding-system 'utf-8)
  (set-selection-coding-system 'utf-8)
  (prefer-coding-system 'utf-8)
#+end_src


.config/doom/config.el
(require 'package)
(setq package-enable-at-startup nil)

(add-to-list 'package-archives
             '("melpa" . "https://melpa.org/packages/"))
(package-initialize)

(setq doom-font  (font-spec :family "JetBrains Mono" :size 17))

(setq doom-theme 'doom-dracula)

(setq display-line-numbers-type t)

(use-package rainbow-mode
  :ensure t
  :init (add-hook 'prog-mode-hook 'rainbow-mode))

(setq display-time-default-load-average nil)

(use-package dashboard
  :ensure t
  :config
    (dashboard-setup-startup-hook)
    (setq dashboard-startup-banner "~/.emacs.d/img/avatar.png")
    (setq dashboard-items '((recents  . 5)
			      (projects . 5)))
    (setq dashboard-banner-logo-title "I am just a coder for fun"))

(setq org-directory "~/org/")

(setq inhibit-startup-message t)

(defun config-visit ()
  (interactive)
  (find-file "~/.config/doom/config.org"))
(global-set-key (kbd "C-c e") 'config-visit)

(defun config-reload ()
  (interactive)
  (org-babel-load-file (expand-file-name "~/.config/doom/config.org")))
(global-set-key (kbd "C-c r") 'config-reload)

(use-package hungry-delete
  :ensure t
  :defer t
  :config (global-hungry-delete-mode))

(use-package atomic-chrome
  :ensure t
  :init
  (setq atomic-chrome-default-major-mode 'markdown-mode)
  (setq atomic-chrome-server-ghost-text-port 4002)
  (setq atomic-chrome-extension-type-list '(atomic-chrome))
  :config
  (atomic-chrome-start-server))

(use-package vterm
  :ensure t
  :init
  (global-set-key (kbd "<s-return>") 'vterm))

(use-package fancy-battery
  :ensure t
  :init
  (fancy-battery-mode 1)
  (setq fancy-battery-show-percentage t))

(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq TeX-save-query nil)
;(setq TeX-PDF-mode t)

(require 'flymake)

(defun flymake-get-tex-args (file-name)
(list "pdflatex"
(list "-file-line-error" "-draftmode" "-interaction=nonstopmode" file-name)))

(add-hook 'LaTeX-mode-hook 'flymake-mode)

(setq ispell-program-name "aspell") ; could be ispell as well, depending on your preferences
(setq ispell-dictionary "english") ; this can obviously be set to any language your spell-checking program supports

(add-hook 'LaTeX-mode-hook 'flyspell-mode)
(add-hook 'LaTeX-mode-hook 'flyspell-buffer)

(defun turn-on-outline-minor-mode ()
(outline-minor-mode 1))

(add-hook 'LaTeX-mode-hook 'turn-on-outline-minor-mode)
(add-hook 'latex-mode-hook 'turn-on-outline-minor-mode)
(setq outline-minor-mode-prefix "\C-c \C-o") ; Or something else

(require 'tex-site)
(autoload 'reftex-mode "reftex" "RefTeX Minor Mode" t)
(autoload 'turn-on-reftex "reftex" "RefTeX Minor Mode" nil)
(autoload 'reftex-citation "reftex-cite" "Make citation" nil)
(autoload 'reftex-index-phrase-mode "reftex-index" "Phrase Mode" t)
(add-hook 'latex-mode-hook 'turn-on-reftex) ; with Emacs latex mode
;; (add-hook 'reftex-load-hook 'imenu-add-menubar-index)
(add-hook 'LaTeX-mode-hook 'turn-on-reftex)

(setq LaTeX-eqnarray-label "eq"
LaTeX-equation-label "eq"
LaTeX-figure-label "fig"
LaTeX-table-label "tab"
LaTeX-myChapter-label "chap"
TeX-auto-save t
TeX-newline-function 'reindent-then-newline-and-indent
TeX-parse-self t
TeX-style-path
'("style/" "auto/"
"/usr/share/emacs/site-lisp/auctex/style/")
LaTeX-section-hook
'(LaTeX-section-heading
LaTeX-section-title
LaTeX-section-toc
LaTeX-section-section
LaTeX-section-label))

(use-package org-latex-impatient
  :defer t
  :hook (org-mode . org-latex-impatient-mode)
  :init
  (setq org-latex-impatient-tex2svg-bin
        ;; location of tex2svg executable
        "~/node_modules/mathjax-node-cli/bin/tex2svg"))

(use-package centaur-tabs
  :ensure t
  :init
  (centaur-tabs-mode 1))

(use-package highlight-indent-guides
  :ensure t
  :defer t
  :hook (prog-mode . highlight-indent-guides-mode)
  :config
  (setq highlight-indent-guides-method 'character)
  (setq highlight-indent-guides-character ?\|)
  (setq highlight-indent-guides-responsive 'top))

(use-package lsp-mode
  :ensure t
  :defer t
  :init
  (setq lsp-keymap-prefix "C-c l")
  :config
  (setq lsp-headerline-breadcrumb-enable nil))

(use-package web-mode
  :ensure t
  :defer t
  :config
  (setq
   web-mode-markup-indent-offset 2
   web-mode-css-indent-offset 2
   web-mode-code-indent-offset 2
   web-mode-style-padding 2
   web-mode-script-padding 2
   web-mode-enable-auto-closing t
   web-mode-enable-auto-opening t
   web-mode-enable-auto-pairing t
   web-mode-enable-auto-indentation t)
  :mode
  (".html$" "*.php$" "*.tsx"))

(use-package emmet-mode
  :ensure t
  :defer t)

(use-package org-bullets
  :ensure t
  :defer t
  :config
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))

(setq org-agenda-files (append
                        (file-expand-wildcards "~/org/*.org")))

(setq locale-coding-system 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(set-selection-coding-system 'utf-8)
(prefer-coding-system 'utf-8)
